src/librustc_passes/rvalue_promotion.rs

```diff
@@ -445,40 +485,141 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
             }
         }
 
-        hir::ExprField(ref expr, _) => {
+        hir::ExprField(ref expr, _ident) => {
+            match *expr {
+                ref expr => { v.visit_expr(&expr) },
+            }
             if let Some(def) = v.tables.expr_ty(expr).ty_adt_def() {
                 if def.is_union() {
                     v.promotable = false
                 }
             }
         }
 
-        hir::ExprBlock(..) |
-        hir::ExprIndex(..) |
-        hir::ExprArray(_) |
-        hir::ExprType(..) |
-        hir::ExprTup(..) => {}
+        hir::ExprBlock(ref box_block, ref _option_label) => {
+            match **box_block {
+                ref block => {
+                    match block.expr {
+                        Some(ref box_expr) => { v.visit_expr(&*box_expr) },
+                        None => {},
+                    } }
+            }
+        }
+
+        hir::ExprIndex(ref lhs, ref rhs) => {
+            v.visit_expr(lhs);
+            v.visit_expr(rhs);
+        }
+
+        hir::ExprArray(ref hirvec) => {
+            for index in hirvec.iter() {
+                match *index {
```

Comment:
you don't need this `match`, it should be fine to call `v.visit_expr(index)` directly, if that fails, use `v.visit_expr(&**index)` to force a deref.

That transformation can be applied to all the `match` statements you used for reference conversions